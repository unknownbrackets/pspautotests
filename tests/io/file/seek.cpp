#include <common.h>
#include <pspiofilemgr.h>

u64 asyncAdapter(int fd, s64 offset, int whence) {
	int result = sceIoLseekAsync(fd, offset, whence);
	if (result < 0) {
		return result;
	}

	SceInt64 res;
	result = sceIoWaitAsync(fd, &res);
	if (result < 0) {
		return result;
	}

	return res;
}

u64 async32Adapter(int fd, int offset, int whence) {
	int result = sceIoLseek32Async(fd, offset, whence);
	if (result < 0) {
		return result;
	}

	SceInt64 res;
	result = sceIoWaitAsync(fd, &res);
	if (result < 0) {
		return result;
	}

	return res;
}

template <typename Func>
void runTests(const char *title, Func seekFunc) {
	int fd = sceIoOpen("Makefile", PSP_O_RDONLY, 0777);
	char fullTitle[256];

	snprintf(fullTitle, sizeof(fullTitle), "%s - files", title);
	checkpointNext(fullTitle);
	checkpoint("  Invalid fd: %016llx", (s64)seekFunc(0xDEADBEEF, 0, SEEK_CUR));
	checkpoint("  0: %016llx", (s64)seekFunc(0, 0, SEEK_CUR));
	checkpoint("  1: %016llx", (s64)seekFunc(1, 0, SEEK_CUR));
	checkpoint("  2: %016llx", (s64)seekFunc(2, 0, SEEK_CUR));
	checkpoint("  Valid fd: %016llx", (s64)seekFunc(fd, 0, SEEK_CUR));

	snprintf(fullTitle, sizeof(fullTitle), "%s - offsets", title);
	checkpointNext(fullTitle);
	checkpoint("  +Negative: %016llx", (s64)seekFunc(fd, -10, SEEK_CUR));
	checkpoint("  +Positive: %016llx", (s64)seekFunc(fd, 10, SEEK_CUR));
	checkpoint("  +Negative again: %016llx", (s64)seekFunc(fd, -10, SEEK_CUR));
	checkpoint("  +Zero: %016llx", (s64)seekFunc(fd, 0, SEEK_CUR));
	checkpoint("  End +1: %016llx", (s64)seekFunc(fd, 1, SEEK_END));
	checkpoint("  End +0: %016llx", (s64)seekFunc(fd, 0, SEEK_END));
	checkpoint("  End +-0x1000: %016llx", (s64)seekFunc(fd, -0x1000, SEEK_END));

	snprintf(fullTitle, sizeof(fullTitle), "%s - whence", title);
	checkpointNext(fullTitle);
	checkpoint("  END: %016llx", (s64)seekFunc(fd, 0, SEEK_END));
	checkpoint("  CUR: %016llx", (s64)seekFunc(fd, 0, SEEK_CUR));
	checkpoint("  SET: %016llx", (s64)seekFunc(fd, 0, SEEK_SET));
	checkpoint("  4: %016llx", (s64)seekFunc(fd, 0, 4));
	checkpoint("  -1: %016llx", (s64)seekFunc(fd, 0, -1));

	sceIoClose(fd);
}

extern "C" int main(int argc, char *argv[]) {
	int fd = sceIoOpen("seek.cpp", PSP_O_RDONLY, 0777);

	runTests("sceIoLseek", sceIoLseek);
	runTests("sceIoLseek32", sceIoLseek32);
	runTests("sceIoLseekAsync", asyncAdapter);
	runTests("sceIoLseek32Async", async32Adapter);

	sceIoClose(fd);
	return 0;
}