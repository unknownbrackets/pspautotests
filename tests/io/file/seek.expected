[x] sceIoLseek - files
[x]   Invalid fd: ffffffff80020323
[x]   0: ffffffff80020325
[x]   1: ffffffff800200d1
[x]   2: ffffffff800200d1
[r]   Valid fd: 0000000000000000

[x] sceIoLseek - offsets
[r]   +Negative: ffffffffffffffff
[r]   +Positive: 000000000000000a
[r]   +Negative again: 0000000000000000
[r]   +Zero: 0000000000000000
[r]   End +1: 000000000000005e
[r]   End +0: 000000000000005d
[r]   End +-0x1000: ffffffffffffffff

[x] sceIoLseek - whence
[r]   END: 000000000000005d
[r]   CUR: 000000000000005d
[r]   SET: 0000000000000000
[x]   4: ffffffff80020324
[x]   -1: ffffffff80020324

[x] sceIoLseek32 - files
[x]   Invalid fd: ffffffff80020323
[x]   0: ffffffff80020325
[x]   1: ffffffff800200d1
[x]   2: ffffffff800200d1
[r]   Valid fd: 0000000000000000

[x] sceIoLseek32 - offsets
[r]   +Negative: ffffffffffffffff
[r]   +Positive: 000000000000000a
[r]   +Negative again: 0000000000000000
[r]   +Zero: 0000000000000000
[r]   End +1: 000000000000005e
[r]   End +0: 000000000000005d
[r]   End +-0x1000: ffffffffffffffff

[x] sceIoLseek32 - whence
[r]   END: 000000000000005d
[r]   CUR: 000000000000005d
[r]   SET: 0000000000000000
[x]   4: ffffffff80020324
[x]   -1: ffffffff80020324

[x] sceIoLseekAsync - files
[x]   Invalid fd: ffffffff80020323
[x]   0: ffffffff80020325
[x]   1: ffffffff800200d1
[x]   2: ffffffff800200d1
[r]   Valid fd: 0000000000000000

[x] sceIoLseekAsync - offsets
[r]   +Negative: ffffffffffffffff
[r]   +Positive: 000000000000000a
[r]   +Negative again: 0000000000000000
[r]   +Zero: 0000000000000000
[r]   End +1: 000000000000005e
[r]   End +0: 000000000000005d
[r]   End +-0x1000: ffffffffffffffff

[x] sceIoLseekAsync - whence
[r]   END: 000000000000005d
[r]   CUR: 000000000000005d
[r]   SET: 0000000000000000
[x]   4: ffffffff80020324
[x]   -1: ffffffff80020324

[x] sceIoLseek32Async - files
[x]   Invalid fd: ffffffff80020323
[x]   0: ffffffff80020325
[x]   1: ffffffff800200d1
[x]   2: ffffffff800200d1
[r]   Valid fd: 0000000000000000

[x] sceIoLseek32Async - offsets
[r]   +Negative: ffffffffffffffff
[r]   +Positive: 000000000000000a
[r]   +Negative again: 0000000000000000
[r]   +Zero: 0000000000000000
[r]   End +1: 000000000000005e
[r]   End +0: 000000000000005d
[r]   End +-0x1000: ffffffffffffffff

[x] sceIoLseek32Async - whence
[r]   END: 000000000000005d
[r]   CUR: 000000000000005d
[r]   SET: 0000000000000000
[x]   4: ffffffff80020324
[x]   -1: ffffffff80020324
